### DISCUSSION TITLE
Why is `PanicException` derived from `BaseException` ?

### DISCUSSION THREAD

**Discussion by rhaps0dy** [PyO3/pyo3#2638] [Questions] [answered]
The docs state that [PanicException derives from BaseException](https://pyo3.rs/main/doc/pyo3/panic/struct.panicexception) so that it makes the interpreter exit. But regular `Exception`s have this effect too.  [User-defined exceptions are meant to be derived from `Exception` according to the Python docs](https://docs.python.org/3/library/exceptions.html). Why did PyO3 decide to inherit from `BaseException` in particular?

I think it'd be better to derive from `Exception`. Some Python code assumes that the only children of BaseException are `SystemExit`, `KeyboardInterrupt`,`GeneratorExit` and `Exception`, which makes sense if all others are meant to derive from `Exception`.

In particular, I'm running into issues with a [Python fuzz testing library](https://github.com/HypothesisWorks/hypothesis) that (rightly) assumes something has gone wrong beyond a test failure when a generic [`BaseException`-derived class is raised](https://github.com/HypothesisWorks/hypothesis/blob/master/hypothesis-python/src/hypothesis/core.py#L543-L556). It makes sense, however, to fuzz-test to find Rust panics. There are easy workarounds for this but it'd be better to fix this for everyone.

**Comment by birkenfeld** [comment, marked as answer]
There is an explanation in the docstring: "Like SystemExit, this exception is derived from BaseException so that it will typically propagate all the way through the stack and cause the Python interpreter to exit." This follows the philosophy that panics are *not* on the same level as exceptions in languages that have them, but something more "severe" that usually leads to an abort if no extraordinary measures are taken (`catch_panic` and `except BaseException`).

It's funny, your link to the hypothesis source shows that pytest defines its own exceptions derived from BaseException. As you can see from that, there is no such universal rule as "all user defined exceptions must derive Exception".

**Reply by rhaps0dy** [reply]
I see, thank you! I'd read that docstring several times, but this helped:

> This follows the philosophy that panics are not on the same level as exceptions in languages that have them, but something more "severe" that usually leads to an abort if no extraordinary measures are taken (catch_panic and except BaseException).

> pytest defines its own exceptions derived from BaseException

Yeah, I'm not sure that's right either, but fair enough.
